### PR TITLE
feature: create app store 

### PR DESCRIPTION
**Pull request by sbender9**
The App Store finds available plugins and web apps from npm and allows the user to install or updated them.

### CODE REVIEW

File: lib/interfaces/appstore.js [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+const _ = require("lodash");
+const fs = require("fs");
+const agent = require("superagent-promise")(require("superagent"), Promise);
+const compareVersions = require("compare-versions");
+const spawn = require("child_process").spawn;
+
+module.exports = function(app) {
+  return {
+    start: function() {
+      app.availablePlugins = [];
```
Comment (tkurki) [inline lib/interfaces/appstore.js]: I am not too crazy about these stashed in the app. You could define them as regular variables before the return statement above and define the functions that need them in the same scope.

File: lib/interfaces/appstore.js [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  npm.on("close", code => {
+    debug("close: " + name);
+    app.pluginsInstalled[name]["code"] = code;
+    app.pluginInstalling = undefined;
+    console.log(`child process exited with code ${code}`);
+
+    if (Object.keys(app.pluginInstallQueue).length) {
+      var next_name = Object.keys(app.pluginInstallQueue)[0];
+      debug("QUEUE: " + next_name);
+      var info = app.pluginInstallQueue[next_name];
```
Comment (tkurki) [inline lib/interfaces/appstore.js]: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/splice

**Comment by tkurki** [comment]
I love the functionality!

Hrmph. Building html by hand is getting quite tiresome, I think we will need something better for the future.

Did you check error handling, like timeout of the installation process and plain errors from installation, like lost net connectivity?

**Comment by sbender9** [comment]
Checked errors and lost connectivity, did not check timeout of install...

**Comment by tkurki** [comment]
Managed to bump into this error condition:
```
stderr: npm
stderr:  WARN update-linked node_modules/tuktuk-chart-plotter needs updating to 0.0.5 from 0.0.1 but we can't, as it's a symlink

child process exited with code 0
```
as I had used npm link to use it.
The ui showed happily _Installed 0.0.5_.

**Comment by sbender9** [comment]
I'm checking the error code from npm , which is ok in the case. That really shouldn't be a warning. I can add a check for linked npms...

**Pull request merged**